design/008-pushsecret-update-policy.md

```diff
@@ -0,0 +1,102 @@
+```yaml
+---
+title: PushSecret Update Policy
+version: v1alpha1
+authors: Moritz Johner
+creation-date: 2023-08-25
+status: draft
+---
+```
+
+# PushSecret Update Policy
+
+## Table of Contents
+
+<!-- toc -->
+// autogen please
+<!-- /toc -->
+
+## Summary
+
+The PushSecret update procedure needs to be configurable so users can control its behaviour.
+
+## Motivation
+
+The motivation is to enhance the configurability of the PushSecret lifecycle within the External Secrets Operator (ESO).
+Currently, there is a need for users to have more control over the behavior of PushSecret when updating values at the secret provider.
+
+The use-case for `updatePolicy` is the following: Users would like to use ESO generate a password and store it in an external store. Once the password is generated and pushed, they want the secret store to become the source of truth, so that the password won't be overridden when ExternalSecret or Secret Resources are deleted and/or regenerated.
+
+More context in #2643.
+
+## Context
+
+When ESO intends to push a secret to the provider, several cases need to be considered for managing the lifecycle:
+
+1. **Ownership of Pre-existing Secret**: Should ESO take ownership over a pre-existing secret?
+2. **Overwriting Existing Value**: Should ESO overwrite an existing value?
+3. **Merging Values**: Should ESO merge structured data?
+
+This design document concerns the second point.
+
+The PushSecret lifecycle consists of these 3 parts:
+
+* **create** the secret on the provider side
+* **update** the secret value every `spec.refreshInterval`
+* **delete** the secret
+
+For each part we can apply a policy. For this design document the **update** part is relevant.
+The following flowchart illustrates the decision-making process:
+
+```mermaid
+flowchart TD
+
+    C{Create} --> |no policy here| F{Update}
+    F -->|updatePolicy=Replace| G[Always overrides the existing value]
+    F -->|updatePolicy=IfNotExists| H[Writes the value only\n if no value exists]
+    G --> F
+    H --> F
+    G --> I{Delete}
+    H --> I{Delete}
+    I --> |deletionPolicy=Delete| J[Deletes the secret]
+    I --> |deletionPolicy=None| K[Retains the existing secret]
+```
+
+## Proposal
+
+The proposed solution is to introduce an `updatePolicy` field within the PushSecret specification. This field allows users to configure the behavior of updates during the PushSecret lifecycle.
+
+By default, the value of updatePolicy is set to `Replace`, meaning that the existing value will always be overridden. Alternatively, users can set the value to `IfNotExists`, which prevents a write operation if a value already exists.
+
```

Comment:
I have one question. If a user specifies the `IfNotExists` policy, even if she updates the Kubernetes Secret data, the change won't be reflected if the secret already exists in upstream, correct? (and that's the point of this new policy)

Reply:
Yes, that's the intention of the policy, which breaks idempotency contracts. That's really my only concern here, is that the users need to know (via documentation) that the state of a secret is 'skipped because it exists, even if it is drifted'

I don't think we should warn for drifts (with kubernetes events), as this is actually expected. So, IMO, this "idempotency break" should be clear on the documentation.